### PR TITLE
add liveness probe proposal

### PR DESCRIPTION
**Pull request by BH4AWS**
### Ⅰ. Describe what this PR does
The proposal of enhanced livenessProbe solution is develop in the Kruise suite. The enhanced livenessProbe solution are composed of two controllers、one webhook and a node probe detection component, 
there are an enhanced livenessProbe webhook, an enhanced livenessProbe map nodePodProbe resource controller, 
an enhanced livenessProbe controller and a node detection component.

### CODE REVIEW

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+
+### Other Notes
+● Pod without OwnerReference, no enhancement ability this solution, degradation to community logic, restarting containers when in failure detection.
+
+● Pod workload need to use the PodUnavailableBudget in kruise, to enable the PubFeature.
+
+● Take precedence over the community logic, both the native livenessProbe configuration and this solution configuration exist
+
+### Release Plan
+● NodePodProbe controller in kruise-daemonset supports TCP/HTTPGET probe besides EXEC checking.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:180]: grpc should also be supported

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+### Implementation
+
+**EnhancedLivenessProbeWebhook** 
+> When a pod created with livenessProbe config with `Annotations: apps.kruise.io/using-enhanced-liveness: true` feature gate, 
+the webhook will admit the pod resource to remove the native livenessProbe and patch the configuration to the pod annotations `apps.kruise.io/container-probe-context: '{...json format for liveness probe with specified contaienrs in pod...}''`.
+
+**EnhancedLivenessProbeMapNodePodProbeController**
+> For using the nodPodProbe controller processing, 
+> the livenessProbe config filed in the pod annotations should be converted to the nodePodProbe custom defined resource in Kruise suite. 
+> This controller can create and update the nodePodProbe resource.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:82]: plz explain in detail about the reconcile logic in pod creating, deleting , exited case

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+
+**EnhancedLivenessProbeController**
+> The overall architecture is shown in Figure 1 as below. 
+> After the enhanced livenessProbe controller watches the state change of node pod probe configured the failed livenessProbe pods, 
+> this decides immediately whether to perform the restart logic using the CRR(container recreate request) or not. 
+
+> In order to take into account the high availability of the application and service protection, 
+> the architecture design is coordinated with the PodUnavailableBudget function in Kruise suite and some anomaly detection algorithms(global switch, black list, global limiting rule, etc.). 
+> These protect the availability of application services before executing the container restart effectively,
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:91]: plz explain in detail about the flow control logic

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+| :white_check_mark:  kruiseDaemon/ProbeDetector detection enabled in bypaas,  strong expansibility                                                                                                        | :no_entry: no                                                                          | 
+| :white_check_mark: the global perspective of the application, the maxUnavailable protection policy and the capactiy building of the resilience protection system(Kruise-manager/EnhancedLivenessProbeController) | :no_entry:  no                                                                         | 
+| :white_check_mark: the architecture is easy to expand, without invasive modification of native components, adapt easily to different scaenarios with strong scalability                                  | :no_entry:  strong coupling, poor scalability                                          | 
+
+
+### Other Notes
+● Pod without OwnerReference, no enhancement ability this solution, degradation to community logic, restarting containers when in failure detection.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:173]: enhanced liveness can be working on pod without ownerreference if the flow control logic does not rely on the workload of pod

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+**EnhancedLivenessProbeController**
+> The overall architecture is shown in Figure 1 as below. 
+> After the enhanced livenessProbe controller watches the state change of node pod probe configured the failed livenessProbe pods, 
+> this decides immediately whether to perform the restart logic using the CRR(container recreate request) or not. 
+
+> In order to take into account the high availability of the application and service protection, 
+> the architecture design is coordinated with the PodUnavailableBudget function in Kruise suite and some anomaly detection algorithms(global switch, black list, global limiting rule, etc.). 
+> These protect the availability of application services before executing the container restart effectively, 
+> so that there is no service available caused by batch application restarts. 
+> For example, all containers fail to be fully detected due to the unexpected reasons such as network or the error probe configuration.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:93]: plz illustrate the container recreate request CR

File: docs/proposals/20240327-enhancedlivenessprobe.md (hunk 1 above)
Comment (BH4AWS) [reply docs/proposals/20240327-enhancedlivenessprobe.md:180]: done

File: docs/proposals/20240327-enhancedlivenessprobe.md (hunk 3 above)
Comment (BH4AWS) [reply docs/proposals/20240327-enhancedlivenessprobe.md:91]: A basic token bucket algorithm framework is released to control the restart frequency of the application with failed livenessProbe pods.

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+● Pod workload need to use the PodUnavailableBudget in kruise, this solution should enable the PubFeature.
+
+● Override the community logic, both the native livenessProbe configuration and this solution configuration exist in the process.
+
+### Release Plan
+● NodePodProbe controller in kruise-daemonset supports TCP/HTTPGET/RPC probe besides EXEC checking.
+
+● EnhancedLivenessProbeWebhook is developed firstly to convert the standard livenessProbe configuration in the special field in pod annotations.
+
+● EnhancedLivenessProbeMapNodePodProbeController is developed to convert the probe config to the nodePodProbe custom resource in kruise.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:296]: Name(EnhancedLivenessProbeMapNodePodProbeController ) is too long , consider rename it to LivenessNodeProbeController

File: docs/proposals/20240327-enhancedlivenessprobe.md [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+| ● kruiseDaemon/ProbeDetector detection enabled in bypaas,  strong expansibility                                                                                                                 | ● no                              | 
+| ● the global perspective of the application, the maxUnavailable protection policy and the capactiy building of the resilience protection system(Kruise-manager/EnhancedLivenessProbeController) | ● no                              | 
+| ● the architecture is easy to expand, without invasive modification of native components, adapt easily to different scaenarios with strong scalability                                          | ● strong coupling, poor scalability | 
+
+
+### Other Notes
+● Since the probe detection is implemented by the nodePodProbe controller in Kruise-daemonset, for the pod without ownerreference, this solution can also take effect on the enhancement ability for restarting containers when in failure livenessProbe status. 
+
+● Pod workload need to use the PodUnavailableBudget in kruise, this solution should enable the PubFeature.
```
Comment (furykerry) [inline docs/proposals/20240327-enhancedlivenessprobe.md:287]: PodUnavailableBudget should not be a required to use enhanced liveness probe

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request merged**